File: lib/features/sales/calculations/sales_order_calculations.dart
// Sales Order calculations (PORTING.md §7).
// 1:1 port of `calculations/salesOrderCalculations.ts`.
//
// Omitted: `getStatusColor` (CSS/AG-Grid color lookup — Flutter uses the
// `StatusBadge` widget).

import '../../../data/models/invoice.dart' show DiscountType;
import '../models/sales_forms.dart' show SOFormItem, flatZeroDiscount;

const List<String> _fieldOrder = [
  'name',
  'quantity',
  'unitPrice',
  'discountValue',
  'taxRate',
];

List<String> getFieldOrder() => _fieldOrder;

String? getNextField(String field) {
  final index = _fieldOrder.indexOf(field);
  return index >= 0 && index + 1 < _fieldOrder.length
      ? _fieldOrder[index + 1]
      : null;
}

int _now() => DateTime.now().millisecondsSinceEpoch;

SOFormItem createEmptyItemRow(int index) {
  return SOFormItem(
    id: _now() + index,
    itemId: '',
    name: '',
    quantity: 1,
    unitPrice: 0,
    taxRate: 0,
    discount: flatZeroDiscount,
  );
}

List<SOFormItem> padItemsToMinimum(List<SOFormItem> items, {int min = 1}) {
  if (items.length >= min) return items;
  final padded = [...items];
  final now = _now();
  for (var i = items.length; i < min; i++) {
    padded.add(SOFormItem(
      id: now + i + 1000,
      itemId: '',
      name: '',
      quantity: 1,
      unitPrice: 0,
      taxRate: 0,
      discount: flatZeroDiscount,
    ));
  }
  return padded;
}

num calculateItemDiscount(SOFormItem item) {    final subtotal = (item.quantity) * (item.unitPrice);
  if (item.discount.type == DiscountType.percentage) {
    return (subtotal * item.discount.value) / 100;
  }
  return item.discount.value;
}

num calculateItemTotal(SOFormItem item) {
  final subtotal = (item.quantity) * (item.unitPrice);
  final discount = calculateItemDiscount(item);
  final afterDiscount = subtotal - discount;
  final taxAmount = (afterDiscount * (item.taxRate)) / 100;
  return afterDiscount + taxAmount;
}

num calculateSubtotal(List<SOFormItem> items) {
  return items.fold<num>(
      0, (sum, item) => sum + (item.quantity) * (item.unitPrice));
}

num calculateDiscount(List<SOFormItem> items) {
  return items.fold<num>(0, (sum, item) => sum + calculateItemDiscount(item));
}

num calculateTax(List<SOFormItem> items) {
  return items.fold<num>(0, (sum, item) {
    final subtotal = (item.quantity) * (item.unitPrice);
    final discount = calculateItemDiscount(item);
    final afterDiscount = subtotal - discount;
    return sum + (afterDiscount * (item.taxRate)) / 100;
  });
}

num calculateTotal(List<SOFormItem> items) {
  return calculateSubtotal(items) - calculateDiscount(items) + calculateTax(items);
}

File: calculations/salesOrderCalculations.ts
/**
 * Sales Order calculations — all pure functions, no React imports.
 */

import type { SOFormItem } from '../types';

const FIELD_ORDER: readonly string[] = ['name', 'quantity', 'unitPrice', 'discountValue', 'taxRate'];

export function getFieldOrder(): readonly string[] {
  return FIELD_ORDER;
}

export function getNextField(field: string): string | undefined {
  const index = FIELD_ORDER.indexOf(field);
  return FIELD_ORDER[index + 1];
}

export function createEmptyItemRow(index: number): SOFormItem {
  return {
    id: Date.now() + index,
    item_id: '',
    name: '',
    quantity: 1,
    unitPrice: 0,
    taxRate: 0,
    discount: { type: 'flat', value: 0 },
  };
}

export function padItemsToMinimum(items: SOFormItem[], min = 1): SOFormItem[] {
  if (items.length >= min) return items;
  const padded = [...items];
  const now = Date.now();
  for (let i = items.length; i < min; i++) {
    padded.push({
      id: now + i + 1000,
      item_id: '',
      name: '',
      quantity: 1,
      unitPrice: 0,
      taxRate: 0,
      discount: { type: 'flat', value: 0 },
    });
  }
  return padded;
}

export function calculateItemDiscount(item: SOFormItem): number {
  const subtotal = (item.quantity || 0) * (item.unitPrice || 0);
  if (item.discount?.type === 'percentage') {
    return (subtotal * (item.discount.value || 0)) / 100;
  }
  return Number(item.discount?.value) || 0;
}

export function calculateItemTotal(item: SOFormItem): number {
  const subtotal = (item.quantity || 0) * (item.unitPrice || 0);
  const discount = calculateItemDiscount(item);
  const afterDiscount = subtotal - discount;
  const taxAmount = (afterDiscount * (item.taxRate || 0)) / 100;
  return afterDiscount + taxAmount;
}

export function calculateSubtotal(items: SOFormItem[]): number {
  return items.reduce((sum, item) => sum + (item.quantity || 0) * (item.unitPrice || 0), 0);
}

export function calculateDiscount(items: SOFormItem[]): number {
  return items.reduce((sum, item) => sum + calculateItemDiscount(item), 0);
}

export function calculateTax(items: SOFormItem[]): number {
  return items.reduce((sum, item) => {
    const subtotal = (item.quantity || 0) * (item.unitPrice || 0);
    const discount = calculateItemDiscount(item);
    const afterDiscount = subtotal - discount;
    return sum + (afterDiscount * (item.taxRate || 0)) / 100;
  }, 0);
}

export function calculateTotal(items: SOFormItem[]): number {
  return calculateSubtotal(items) - calculateDiscount(items) + calculateTax(items);
}

export { getStatusColor } from './statusColors';
